File: py/Film.py
class Film:
    def __init__(self, kode:str, judul:str, director:str, genre:str, tahun:int):
        self.__kode = str(kode) #kode unik
        self.__judul =  str(judul)
        self.__director = str(director)
        self.__genre = str(genre) #genrenya satu genre utama aja
        self.__tahun = int(tahun)

    daftarfilm = [] #buat array

    #ini getter haha
    def getcode(self) -> str:
        return self.__kode
    
    def getJewdul(self) -> str: #judul
        return self.__judul

    def getDirector(self) -> str: #direktor
        return self.__director

    def getGenre(self) -> str: #genre
        return self.__genre

    def getYear(self) -> int: #tahun
        return self.__tahun

    #setter, urutannya sama dah kayak getter
    def setcode(self, kode:str) -> None:
        self.__kode = str(kode)
    
    def setJew(self, judul:str) -> None:
        self.__judul = str(judul)

    def setDir(self, director:str) -> None:
        self.__director = str(director)

    def setGenre(self, genre:str) -> None:
        self.__genre = str(genre)

    def setYr(self, tahun:int) -> None:
        self.__tahun = int(tahun)

    #CRUD
    #display
    @classmethod
    def displayArr(cls):
        if not cls.daftarfilm:
            print("Daftar film kosong.")
            return

        for film in cls.daftarfilm:
            print("---------------------------")
            print(f"{film.getcode()}")
            print(f"Judul    : {film.getJewdul()}")
            print(f"Sutradara: {film.getDirector()}")
            print(f"Genre    : {film.getGenre()}")
            print(f"Tahun    : {film.getYear()}")
        print("---------------------------")

    #ini search berdasarkan kode  
    @classmethod  
    def searchKode(cls):
        print("--- Search Bar ---")
        code = input("Cari kode film(ex -> F001): ").upper()

        found = False
        for film in cls.daftarfilm:
            if code == film.getcode():
                print(f"{film.getcode()}")
                print(f"Judul    : {film.getJewdul()}")
                print(f"Sutradara: {film.getDirector()}")
                print(f"Genre    : {film.getGenre()}")
                print(f"Tahun    : {film.getYear()}")
                found = True
                break #kalo ketemu, loop langsung berenti, karena kode sifatnya unique
        if not found :
            print("gaada mas")

    @classmethod
    def searchF(cls):
        flm = input("Cari judul film: ")

        found = False
        for film in cls.daftarfilm:
            if flm == film.getJewdul:
                print(f"{film.getcode()}")
                print(f"Judul    : {film.getJewdul()}")
                print(f"Sutradara: {film.getDirector()}")
                print(f"Genre    : {film.getGenre()}")
                print(f"Tahun    : {film.getYear()}")
                found = True #gaada break, bisa aja sama judul tapi remake atau apa gitu
        if not found :
            print("gaada mas")

    # Tambah Data (Create)
    @classmethod
    def addFilm(cls):
        print("--- Tambah Film Baru ---")
        kode = input("Masukkan Kode (ex: F001): ").strip().upper()

        #biar kode gaada sama
        for film in cls.daftarfilm:
            if film.getcode() == kode:
                print("Error: Kode film sudah terdaftar!")
                return

        #input
        judul = input("Masukkan Judul Film: ").strip()
        director = input("Masukkan Sutradara: ").strip()
        genre = input("Masukkan Genre: ").strip()

        try:
            tahun = int(input("Masukkan Tahun Rilis: "))
        except ValueError: #error handling biar tahun sesuai format intejer
            print("Error: Tahunnya angka mas")
            return

        #tambah ke array
        film_baru = cls(kode, judul, director, genre, tahun)
        cls.daftarfilm.append(film_baru)
        print("!! Homre. Film udah ditambah yah !!")

    # Hapus Data (Delete)
    @classmethod
    def deleteFilm(cls):
        print("--- Hapus Film ---")
        kode = input("Masukkan Kode film yang ingin dihapus: ").strip().upper()

        for i, film in enumerate(cls.daftarfilm):
            if film.getcode() == kode:
                cls.daftarfilm.pop(i)
                print(f"!! Oke, Film dengan kode {kode} udah dihapus !!")
                return
        print("gaada mas")

    # Update Data
    @classmethod
    def updateFilm(cls):
        print("--- Update Data Film ---")
        kode = input("Masukkan Kode film yang ingin diupdate: ").strip().upper()

        for film in cls.daftarfilm:
            if film.getcode() == kode:
                #search
                print("Data lama ditemukan:")
                print(f"Judul    : {film.getJewdul()}")
                print(f"Sutradara: {film.getDirector()}")
                print(f"Genre    : {film.getGenre()}")
                print(f"Tahun    : {film.getYear()}")
                print("---------------------------")
                print("Masukkan data baru (kosongkan / tekan Enter jika tidak ingin diubah):")

                judul_baru = input(f"Judul baru [{film.getJewdul()}]: ").strip()
                if judul_baru:
                    film.setJew(judul_baru)

                director_baru = input(f"Sutradara baru [{film.getDirector()}]: ").strip()
                if director_baru:
                    film.setDir(director_baru)

                genre_baru = input(f"Genre baru [{film.getGenre()}]: ").strip()
                if genre_baru:
                    film.setGenre(genre_baru)

                tahun_baru = input(f"Tahun baru [{film.getYear()}]: ").strip()
                if tahun_baru:
                    film.setYr(tahun_baru)

                print("!! Homre. Data film berhasil diupdate !!")
                return
        print("gaada mas")

File: cpp/Film.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Film {
private:
    string kode; //kode unik
    string judul;
    string director;
    string genre; //genrenya satu genre utama aja
    int tahun;

public:
    static vector<Film> daftarfilm; //buat array

    // constructor & destructor
    Film() {
        this->kode = "";
        this->judul = "";
        this->director = "";
        this->genre = "";
        this->tahun = 0;
    }

    Film(string kode, string judul, string director, string genre, int tahun) {
        this->kode = kode;
        this->judul = judul;
        this->director = director;
        this->genre = genre;
        this->tahun = tahun;
    }

    ~Film() {}

    //ini getter haha
    string getcode() {
        return this->kode;
    }

    string getJewdul() { //judul
        return this->judul;
    }

    string getDirector() { //direktor
        return this->director;
    }

    string getGenre() { //genre
        return this->genre;
    }

    int getYear() { //tahun
        return this->tahun;
    }

    //setter, urutannya sama dah kayak getter
    void setcode(string kode) {
        this->kode = kode;
    }

    void setJew(string judul) {
        this->judul = judul;
    }

    void setDir(string director) {
        this->director = director;
    }

    void setGenre(string genre) {
        this->genre = genre;
    }

    void setYr(int tahun) {
        this->tahun = tahun;
    }

    //CRUD
    //display
    static void displayArr() {
        if (daftarfilm.empty()) {
            cout << "Daftar film kosong." << endl;
            return;
        }

        for (int i = 0; i < daftarfilm.size(); i++) {
            cout << "---------------------------" << endl;
            cout << daftarfilm[i].getcode() << endl;
            cout << "Judul    : " << daftarfilm[i].getJewdul() << endl;
            cout << "Sutradara: " << daftarfilm[i].getDirector() << endl;
            cout << "Genre    : " << daftarfilm[i].getGenre() << endl;
            cout << "Tahun    : " << daftarfilm[i].getYear() << endl;
        }
        cout << "---------------------------" << endl;
    }

    //ini search berdasarkan kode  
    static void searchKode() {
        cout << "--- Search Bar ---" << endl;
        cout << "Cari kode film(ex -> F001): ";
        string code;
        getline(cin, code);
        for (int i = 0; i < code.length(); i++) {
            code[i] = toupper(code[i]);
        }

        bool found = false;
        for (int i = 0; i < daftarfilm.size(); i++) {
            if (code == daftarfilm[i].getcode()) {
                cout << daftarfilm[i].getcode() << endl;
                cout << "Judul    : " << daftarfilm[i].getJewdul() << endl;
                cout << "Sutradara: " << daftarfilm[i].getDirector() << endl;
                cout << "Genre    : " << daftarfilm[i].getGenre() << endl;
                cout << "Tahun    : " << daftarfilm[i].getYear() << endl;
                found = true;
                break; //kalo ketemu, loop langsung berenti, karena kode sifatnya unique
            }
        }
        if (!found) {
            cout << "gaada mas" << endl;
        }
    }

    static void searchF() {
        cout << "Cari judul film: ";
        string flm;
        getline(cin, flm);

        bool found = false;
        for (int i = 0; i < daftarfilm.size(); i++) {
            if (flm == daftarfilm[i].getJewdul()) {
                cout << daftarfilm[i].getcode() << endl;
                cout << "Judul    : " << daftarfilm[i].getJewdul() << endl;
                cout << "Sutradara: " << daftarfilm[i].getDirector() << endl;
                cout << "Genre    : " << daftarfilm[i].getGenre() << endl;
                cout << "Tahun    : " << daftarfilm[i].getYear() << endl;
                found = true; //gaada break, bisa aja sama judul tapi remake atau apa gitu
            }
        }
        if (!found) {
            cout << "gaada mas" << endl;
        }
    }

    // Tambah Data (Create)
    static void addFilm() {
        cout << "--- Tambah Film Baru ---" << endl;
        cout << "Masukkan Kode (ex: F001): ";
        string kode;
        getline(cin, kode);
        for (int i = 0; i < kode.length(); i++) {
            kode[i] = toupper(kode[i]);
        }

        //biar kode gaada sama
        for (int i = 0; i < daftarfilm.size(); i++) {
            if (daftarfilm[i].getcode() == kode) {
                cout << "Error: Kode film sudah terdaftar!" << endl;
                return;
            }
        }

        //input
        cout << "Masukkan Judul Film: ";
        string judul;
        getline(cin, judul);

        cout << "Masukkan Sutradara: ";
        string director;
        getline(cin, director);

        cout << "Masukkan Genre: ";
        string genre;
        getline(cin, genre);

        int tahun;
        cout << "Masukkan Tahun Rilis: ";
        if (!(cin >> tahun)) { //error handling biar tahun sesuai format intejer
            cout << "Error: Tahunnya angka mas" << endl;
            cin.clear();
            cin.ignore(1000, '\n');
            return;
        }
        cin.ignore();

        //tambah ke array
        Film film_baru(kode, judul, director, genre, tahun);
        daftarfilm.push_back(film_baru);
        cout << "!! Homre. Film udah ditambah yah !!" << endl;
    }

    // Hapus Data (Delete)
    static void deleteFilm() {
        cout << "--- Hapus Film ---" << endl;
        cout << "Masukkan Kode film yang ingin dihapus: ";
        string kode;
        getline(cin, kode);
        for (int i = 0; i < kode.length(); i++) {
            kode[i] = toupper(kode[i]);
        }

        for (int i = 0; i < daftarfilm.size(); i++) {
            if (daftarfilm[i].getcode() == kode) {
                daftarfilm.erase(daftarfilm.begin() + i);
                cout << "!! Oke, Film dengan kode " << kode << " udah dihapus !!" << endl;
                return;
            }
        }
        cout << "gaada mas" << endl;
    }

    // Update Data
    static void updateFilm() {
        cout << "--- Update Data Film ---" << endl;
        cout << "Masukkan Kode film yang ingin diupdate: ";
        string kode;
        getline(cin, kode);
        for (int i = 0; i < kode.length(); i++) {
            kode[i] = toupper(kode[i]);
        }

        for (int i = 0; i < daftarfilm.size(); i++) {
            if (daftarfilm[i].getcode() == kode) {
                //search
                cout << "Data lama ditemukan:" << endl;
                cout << "Judul    : " << daftarfilm[i].getJewdul() << endl;
                cout << "Sutradara: " << daftarfilm[i].getDirector() << endl;
                cout << "Genre    : " << daftarfilm[i].getGenre() << endl;
                cout << "Tahun    : " << daftarfilm[i].getYear() << endl;
                cout << "---------------------------" << endl;
                cout << "Masukkan data baru (kosongkan / tekan Enter jika tidak ingin diubah):" << endl;

                cout << "Judul baru [" << daftarfilm[i].getJewdul() << "]: ";
                string judul_baru;
                getline(cin, judul_baru);
                if (!judul_baru.empty()) {
                    daftarfilm[i].setJew(judul_baru);
                }

                cout << "Sutradara baru [" << daftarfilm[i].getDirector() << "]: ";
                string director_baru;
                getline(cin, director_baru);
                if (!director_baru.empty()) {
                    daftarfilm[i].setDir(director_baru);
                }

                cout << "Genre baru [" << daftarfilm[i].getGenre() << "]: ";
                string genre_baru;
                getline(cin, genre_baru);
                if (!genre_baru.empty()) {
                    daftarfilm[i].setGenre(genre_baru);
                }

                cout << "Tahun baru [" << daftarfilm[i].getYear() << "]: ";
                string tahun_baru;
                getline(cin, tahun_baru);
                if (!tahun_baru.empty()) {
                    daftarfilm[i].setYr(stoi(tahun_baru));
                }

                cout << "!! Homre. Data film berhasil diupdate !!" << endl;
                return;
            }
        }
        cout << "gaada mas" << endl;
    }
};

// inisialisasi array static
vector<Film> Film::daftarfilm;